rust/src/io/reader.rs

```diff
@@ -60,7 +77,196 @@ pub async fn read_manifest(object_store: &ObjectStore, path: &Path) -> Result<Ma
+                    page_info.position,
+                    page_info.length,
+                )?)
+            };
+        }
+
+        let decoder: Box<dyn Decoder> = match field.data_type() {
+            Int8 => create_plain_decoder!(Int8Type),
+            Int16 => create_plain_decoder!(Int16Type),
+            Int32 => create_plain_decoder!(Int32Type),
+            Int64 => create_plain_decoder!(Int64Type),
+            UInt8 => create_plain_decoder!(UInt8Type),
+            UInt16 => create_plain_decoder!(UInt16Type),
+            UInt32 => create_plain_decoder!(UInt32Type),
+            UInt64 => create_plain_decoder!(UInt64Type),
+            Float16 => create_plain_decoder!(Float16Type),
+            Float32 => create_plain_decoder!(Float32Type),
+            Float64 => create_plain_decoder!(Float64Type),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("Unsupport primitive type: {}", field.data_type()),
+                ))
+            }
+        };
+        decoder.decode().await
+    }
+
+    async fn read_binary_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let column = field.id;
+        let page_info = self.page_table.get(column, batch_id).ok_or(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "field: {} batch {} does not exist in page table",
+                field, batch_id
+            ),
+        ))?;
+
+        use DataType::*;
+        let decoder: Box<dyn Decoder> = match field.data_type() {
+            Utf8 => Box::new(BinaryDecoder::<Utf8Type>::new(
+                &self.object_reader,
+                page_info.position,
+                page_info.length,
+            )),
+            Binary => Box::new(BinaryDecoder::<BinaryType>::new(
+                &self.object_reader,
+                page_info.position,
+                page_info.length,
+            )),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("Unsupported binary type: {}", field.data_type()),
+                ))
+            }
+        };
+        decoder.decode().await
+    }
+
+    /// Read an array of the batch.
+    async fn read_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        if field.data_type().is_numeric() {
+            return self.read_primitive_array(field, batch_id).await;
+        } else if matches!(
+            field.data_type(),
+            DataType::Utf8 | DataType::LargeUtf8 | DataType::Binary | DataType::LargeBinary
+        ) {
+            self.read_binary_array(field, batch_id).await
+        } else {
+            println!("Reading: {}", field);
+            let page_info = self.page_table.get(field.id, batch_id).unwrap();
+            Ok(Arc::new(Int32Array::from(
+                (0..page_info.length as i32).collect::<Vec<_>>(),
+            )))
```

Comment:
just checking my understanding - nested array handling would just be additional branches here right?

Reply:
Yes this is just a place holder, we need to handle struct / list here, which will be a composite call to call read_array() for each subfield.